mitmproxy/http2.py

```diff
@@ -0,0 +1,946 @@
+from mitmproxy import flow
+from mitmproxy import controller
+from mitmproxy import viewitem
+from mitmproxy.coretypes import callbackdict
+import time
+from typing import List
+import typing
+
+import h2.events
+from h2.errors import ErrorCodes
+from h2.settings import SettingCodes, ChangedSetting
+from hpack.hpack import HeaderTuple
+
+
+class _EndStreamFrame():
+
+    """
+        Used for all frame type which can have the "end stream" flags
+    """
+
+    def __init__(self, end_stream):
+        self._end_stream: bool = end_stream
+
+    @property
+    def end_stream(self):
+        return self._end_stream
+
+    @end_stream.setter
+    def end_stream(self, end_stream: bool):
+        if self._events:
+            if end_stream and not self._end_stream:
+                new_event = h2.events.StreamEnded()
+                new_event.stream_id = self._stream_id
+                for event in self._events:
+                    if hasattr(event, "stream_ended"):
+                        event.stream_ended = new_event
+                self._events.append(new_event)
+            elif not end_stream:
+                for event in self._events:
+                    if hasattr(event, "stream_ended"):
+                        event.stream_ended = None
+                    elif isinstance(event, h2.events.StreamEnded):
+                        self._events.remove(event)
+        self._end_stream = end_stream
+
+
+class _PriorityFrame():
```

Comment:
Instead of re-implementing all these frame classes, would it be possible to reuse the class hierarchy from our hyperframe dependency? https://github.com/python-hyper/hyperframe/blob/82abce835659f46d3944c65dce62d584152d4aab/hyperframe/frame.py